pkg/cri/relay/runtime-service.go

```diff
@@ -206,3 +221,80 @@ func (r *relay) RuntimeConfig(ctx context.Context, req *criv1.RuntimeConfigReque
 	r.dump("RuntimeConfig", req)
 	return r.client.RuntimeConfig(ctx, req)
 }
+
+const (
+	eventRelayTimeout = 1 * time.Second
+)
+
+func (r *relay) addEventServer(req *criv1.GetEventsRequest) chan *criv1.ContainerEventResponse {
+	r.Lock()
+	defer r.Unlock()
+
+	evtC := make(chan *criv1.ContainerEventResponse, 128)
+	r.evtChans[req] = evtC
+
+	return evtC
+}
+
+func (r *relay) delEventServer(req *criv1.GetEventsRequest) chan *criv1.ContainerEventResponse {
+	r.Lock()
+	defer r.Unlock()
+
+	evtC := r.evtChans[req]
+	delete(r.evtChans, req)
+
+	return evtC
+}
+
+func (r *relay) startEventRelay(req *criv1.GetEventsRequest) error {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.evtClient != nil {
+		return nil
+	}
+
+	c, err := r.client.GetContainerEvents(context.Background(), req)
+	if err != nil {
+		return fmt.Errorf("failed to create container event client: %w", err)
+	}
+
+	r.evtClient = c
+	go r.relayEvents()
```

Comment:
I don't know if we need to support multiple `GetContainerEvents()` with different `req`s on the same relay that might be calling two `startEventRelay()`s concurrently.

If we don't, maybe we should not allow overwriting `r.evtClient` and launching the relaying goroutine.

If we do, we might want to handle relaying here a bit differently.

Do we need separate goroutine If we need a separate goroutine `r.relayEvents()` to receive events from each client? Then it would be safer to call `go r.relayEvents(c)` and try to manage without `r.evtClient` that get overwritten on line 262.

Reply:
But this shouldn't happen because the first one added `r.evtClient` and the second req bails out before starting the goroutine. Or did I read this right?

Reply:
Yes, of course. Thanks @marquiz. I'm dazzled by my blindness.